src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+    NOTJOINED = 2
+    FAULT = 3
+
+
+class SRV(enum.Enum):
+    DOMAINCONTROLLER = '_ldap._tcp.dc._msdcs.'
+    FORESTGLOBALCATALOG = '_ldap._tcp.gc._msdcs.'
+    GLOBALCATALOG = '_gc._tcp.'
+    KERBEROS = '_kerberos._tcp.'
+    KERBEROSDOMAINCONTROLLER = '_kerberos._tcp.dc._msdcs.'
+    KPASSWD = '_kpasswd._tcp.'
+    LDAP = '_ldap._tcp.'
+    PDC = '_ldap._tcp.pdc._msdcs.'
+
+
+class SSL(enum.Enum):
+    NOSSL = 'off'
+    USESSL = 'on'
+    USETLS = 'start_tls'
+
+
+class ActiveDirectory_DNS(object):
+    """
+    :get_n_working_servers: often only a few working servers are needed and not the whole
+    list available on the domain. This takes the SRV record type and number of servers to get
+    as arguments.
+    """
+    def __init__(self, **kwargs):
+        super(ActiveDirectory_DNS, self).__init__()
+        self.ad = kwargs.get('conf') 
+        self.logger = kwargs.get('logger')
+        return
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        if typ is not None:
+            raise
+
+    def _get_SRV_records(self, host, dns_timeout):
+        """
+        Set resolver timeout to 1/3 of the lifetime. The timeout defines
+        how long to wait before moving on to the next nameserver in resolv.conf
+        """
+        srv_records = []
+
+        if not host:
+            return srv_records
+
+        r = resolver.Resolver()
+        r.lifetime = dns_timeout
+        r.timeout = r.lifetime / 3
+
+        try:
+
+            answers = r.query(host, 'SRV')
+            srv_records = sorted(
+                answers,
+                key=lambda a: (int(a.priority), int(a.weight))
+            )
+
+        except Exception as e:
+            srv_records = []
+
+        return srv_records
+
+    def port_is_listening(self, host, port, timeout=1):
+        ret = False
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        if timeout:
+            s.settimeout(timeout)
+
+        try:
+            s.connect((host, port))
+            ret = True
+
+        except Exception as e:
+            s.close()
+            raise CallError(e)
+
+        s.close()
+        return ret
+
+    def _get_servers(self, srv_prefix):
+        """
+        We will first try fo find servers based on our AD site. If we don't find
+        a server in our site, then we populate list for whole domain. Ticket #27584
+        Domain Controllers, Forest Global Catalog Servers, and Kerberos Domain Controllers
+        need the site information placed before the 'msdcs' component of the host entry.t
+        """
+        servers = []
+        if not self.ad['domainname']:
+            return servers 
+
+        if self.ad['site'] and self.ad['site'] != 'Default-First-Site-Name':
+            if 'msdcs' in srv_prefix.value: 
+                parts = srv_prefix.value.split('.')
+                srv = '.'.join([parts[0],parts[1]])
+                msdcs = '.'.join([parts[2], parts[3]])
```

Comment:
Is there a guarantee we always get 4 parts?

Reply:
Yes.